addons/knowledge/models/knowledge_article.py

```diff
@@ -13,6 +13,51 @@
 from odoo.osv import expression
 from odoo.tools import get_lang, formataddr
 
+class ArticleFavourites(models.Model):
+    _name = 'knowledge.article.favourite'
+    _description = 'Favourite Articles'
+    _table = 'knowledge_article_favourite_user_rel'
+    _order = 'sequence asc'
+
+    article_id = fields.Many2one('knowledge.article', 'Article', required=True, ondelete='cascade')
+    user_id = fields.Many2one('res.users', 'User', index=True, required=True, ondelete='cascade')
+    sequence = fields.Integer(default=1)
+
+    def create(self, vals):
+        res = super(ArticleFavourites, self).create(vals)
+        res.reorder_favourite(res.article_id.id)
+        return res
+
+    def _set_sequence(self, article_ids):
+        for article_id in article_ids:
+            self.reorder_favourite(article_id)
+
+    def reorder_favourite(self, article_id, before_article_id=False):
+        user_articles = self.sudo().search([('user_id', '=', self.env.user.id)])
+        article = user_articles.filtered(lambda a: a.article_id.id == article_id)
+        if not article:
+            raise UserError(_("You don't have this article in your favourites."))
+        if not article.sequence:
+            article.sequence = 0
+        if not before_article_id:
+            max_sequence = user_articles.search([('user_id', '=', self.env.user.id)], order='sequence desc', limit=1).sequence or 0
+            # FIXME: values of sequence will keep increasing forever
+            article.sequence = max_sequence + 1
+            return True
+        before_article = user_articles.filtered(lambda a: a.article_id.id == before_article_id)
+        if not before_article:
+            raise UserError(_("You don't have this article in your favourites."))
+        new_sequence = before_article.sequence
+        to_update = user_articles.search([('sequence', '>=', new_sequence)])
+        for article in to_update:
+            article.sequence += 1
+        article.sequence = new_sequence
```

Comment:
You increase every article by 1, then you change one of them to put it at new sequence. which means you set twice the sequence of the same article.
```
new_sequence = before_article.sequence
article.sequence = new_sequence
to_update = user_articles.search([('sequence', '>=', new_sequence), ('article_id', '!=', article.id)])
```